File: PythonFolders/network_stats.py
from flask import Flask, jsonify, request
from flask_cors import CORS
import psutil
import time
from pysnmp.hlapi import *
import socket
import socket
import subprocess

app = Flask(__name__)
CORS(app)

# Simulated database for devices with online status
devices = {}

def get_network_stats(ip_address):
    stats = psutil.net_io_counters()
    host = socket.gethostname()
    # Ping the provided IP address to get latency
    latency = subprocess.check_output(f"ping -n 1 {ip_address} | findstr Average", shell=True)
    latency = latency.decode('utf-8').strip().split('=')[-1].split('ms')[0]
    return {
        'bytes_sent': stats.bytes_sent,
        'bytes_recv': stats.bytes_recv,
        'packets_sent': stats.packets_sent,
        'packets_recv': stats.packets_recv,
        'timestamp': time.strftime('%Y-%m-%d %H:%M:%S'),
        'host': host,
        'latency': latency
    }

# Route to fetch network stats
@app.route('/stats', methods=['GET'])
def stats():
    return jsonify(get_network_stats("localhost"))

from pythonping import ping

def is_device_online(ip_address):
    try:
        response = ping(ip_address, count=1, timeout=2)
        return True if response.success() else False
    except Exception:
        return False

# Route to fetch device info
@app.route('/device-info', methods=['POST'])
def device_info():
    data = request.get_json()
    ip_address = data.get('ip_address')
    return jsonify(get_device_info(ip_address))

# Route to add a new device
@app.route('/add-device', methods=['POST'])
def add_device():
    data = request.get_json()
    ip_address = data.get('ip_address')

    if ip_address in devices:
        return jsonify({'message': 'Device already exists'}), 400

    # Check if the device is online
    if is_device_online(ip_address):
        # Get the network stats of the device
        network_stats = get_network_stats(ip_address)
        latency = network_stats['latency']

        # Add the device to the database with the initial online status and latency
        devices[ip_address] = {
            'ip_address': ip_address,
            'added_on': time.strftime('%Y-%m-%d %H:%M:%S'),
            'status': 'online',
            'latency': latency
        }
        return jsonify({'message': 'Device added successfully'})
    else:
        return jsonify({'message': 'Device is not reachable or does not respond'}), 400
# Route to remove a device
@app.route('/remove-device', methods=['POST'])
def remove_device():
    data = request.get_json()
    ip_address = data.get('ip_address')

    if ip_address not in devices:
        return jsonify({'message': 'Device does not exist'}), 400

    # Simulate removing device from database
    del devices[ip_address]

    return jsonify({'message': 'Device removed successfully'})

@app.route('/devices', methods=['GET'])
def devices_list():
    updated_devices = []
    for ip_address, device_info in devices.items():
        if is_device_online(ip_address):
            device_info['status'] = 'online'
            network_stats = get_network_stats(ip_address)
            latency = network_stats['latency']
            device_info['latency'] = latency
        else:
            device_info['status'] = 'offline'
            device_info['latency'] = None


        updated_devices.append(device_info)
    return jsonify(updated_devices)

if __name__ == "__main__":
    app.run(port=5001)

File: Client/network-management-dashboard/src/NetworkStats.js
import React from 'react';
import { useQuery, useMutation, gql } from '@apollo/client';

const GET_NETWORK_STATS = gql`
  query GetNetworkStats {
    getNetworkStats {
      bytes_sent
      bytes_recv
      packets_sent
      packets_recv
      timestamp
      host
      latency
    }
  }
`;

const GET_DEVICES = gql`
  query GetDevices {
    getDevices {
      ip_address
      added_on
      status
      latency
    }
  }
`;

const ADD_DEVICE = gql`
  mutation AddDevice($ip_address: String!) {
    addDevice(ip_address: $ip_address)
  }
`;

const REMOVE_DEVICE = gql`
  mutation RemoveDevice($ip_address: String!) {
    removeDevice(ip_address: $ip_address)
  }
`;

const NetworkStats = () => {
  const { loading, error, data } = useQuery(GET_NETWORK_STATS);

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error :(</p>;

  return (
    <div>
      <h2>Network Stats</h2>
      <p>Bytes Sent: {data.getNetworkStats.bytes_sent}</p>
      <p>Bytes Received: {data.getNetworkStats.bytes_recv}</p>
      <p>Packets Sent: {data.getNetworkStats.packets_sent}</p>
      <p>Packets Received: {data.getNetworkStats.packets_recv}</p>
      <p>Timestamp: {data.getNetworkStats.timestamp}</p>
      <p>Host: {data.getNetworkStats.host}</p>
      <p>Latency: {data.getNetworkStats.latency}</p>
    </div>
  );
};

const Devices = () => {
  const { loading, error, data } = useQuery(GET_DEVICES);
  const [addDevice] = useMutation(ADD_DEVICE);
  const [removeDevice] = useMutation(REMOVE_DEVICE);

  const handleAddDevice = (ip) => {
    addDevice({ variables: { ip_address: ip } });
  };

  const handleRemoveDevice = (ip) => {
    removeDevice({ variables: { ip_address: ip } });
  };

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error :(</p>;

  return (
    <div>
      <h2>Devices</h2>
      <ul>
        {data.getDevices.map((device) => (
          <li key={device.ip_address}>
            {device.ip_address} - {device.status} - {device.latency}
            <button onClick={() => handleRemoveDevice(device.ip_address)}>Remove</button>
          </li>
        ))}
      </ul>
      <button onClick={() => handleAddDevice('192.168.1.100')}>Add Device</button>
    </div>
  );
};

const App = () => (
  <div>
    <h1>Network Management Dashboard</h1>
    <NetworkStats />
    <Devices />
  </div>
);

export default App;
